### PR TITLE
FM2-100: Return an empty list when searching by an invalid gender code

### PR DESCRIPTION
**Pull request by CaptainDaVinci**
## Description of what I changed

Return an empty list when searching by an invalid gender code.

## Issue I worked on

see https://issues.openmrs.org/browse/FM2-100

## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**.

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

File: api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirPatientDaoImplTest.java [hunk 1]
```diff
@@ -271,7 +271,7 @@ public void searchForPatients_shouldReturnPatientsByGender() {
 	@Test
 	public void searchForPatients_shouldReturnEmptyCollectionWhenGenderNotMatched() {
 		Collection<Patient> results = dao.searchForPatients(null, null, null, null,
-		    new TokenOrListParam().add(new TokenParam("other")), null, null, null, null, null, null, null, null);
+		    new TokenOrListParam().add(new TokenParam("wrong-gender")), null, null, null, null, null, null, null, null);
```
Comment (ibacher) [inline api/src/test/java/org/openmrs/module/fhir2/api/dao/impl/FhirPatientDaoImplTest.java]: We should probably extract this out into a constant

File: api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java [hunk 2]
```diff
@@ -440,7 +440,7 @@ protected boolean containsAlias(Iterator<CriteriaImpl.Subcriteria> subcriteriaIt
 			}
```
Comment (ibacher) [inline api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java]: We should also address the `Optional.empty()` above this line.

Comment (CaptainDaVinci) [reply api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java]: I have a couple of questions here, 

1. The code `OTHER` has a mapping to `'O'`, as shown in the `GenderTranslatorImplTest`. Keeping that in mind, should we change the Criterion to match `'O'` specifically? Because the code `UNKNOWN` too has a mapping to `'U'`.

2. Should `NULL` match any resource or those which have gender set to empty or null (if that's even possible)?

Comment (ibacher) [reply api/src/main/java/org/openmrs/module/fhir2/api/dao/impl/BaseDaoImpl.java]: So remember what we're mapping to here: we're querying the OpenMRS database for people matching whatever gender. In OpenMRS, there are really just three possible gender codes `M`, `F` and `null`.  Maybe a better way to support this is to map all of `OTHER`, `UNKNOWN`, and `NULL` to `gender is null`. We should probably fix the gender translator to reflect this as well. I suppose it should just return `M` for "male", `F` for "female" and `null` for "unknown".

**Comment by ibacher** [comment]
@tendomart That was my fault! @CaptainDaVinci please rebase this commit on the current master.

**Comment by tendomart** [comment]
Ok alright thanks for the info @ibacher

**Comment by ibacher** [comment]
@CaptainDaVinci This looks good so far. I think we need tests for the genders `OTHER`, `UNKNOWN` and `NULL` at the DAO level to ensure that they behave properly and this can be merged. Thanks for your work on this!

**Comment by CaptainDaVinci** [comment]
@ibacher I made a few more changes to the `handleGender` method. 

1. The gender code `OTHER` is mapped to resources with gender set to `nulll`.
2. Looking into the `AdministrativeGender.fromCode` method, the code `NULL` is not mapped to `null` value, hence an explicit check is added.
3.  The tests for `PersonDaoImpl` and `PatientDao` now check for if every result item has the matching gender.

P.S: Additional patient data may need to be added to the core data set, to test for `UNKNOWN`, `NULL` and `OTHER` gender code.
